### ISSUE TITLE
[Typescript support]: compilation breaks with exactOptionalPropertyTypes enabled in userland

### ISSUE THREAD

**Issue by notaphplover** [facebook/docusaurus#6295]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

Typescript configurations extending `@tsconfig/docusaurus/tsconfig.json` breaks on strict checks.

As ts developer I expect ts configs extending `@tsconfig/docusaurus/tsconfig.json` work even if strict checks are enabled.

### Steps to reproduce

1. Create a fresh docusaurus typescript project `npx create-docusaurus@latest sample-project`.
2. Navigate to the project root directory `cd sample-project`
3. Update `tsconfig.json` file:

```json
{
  "extends": "@tsconfig/docusaurus/tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "exactOptionalPropertyTypes": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["ES2021"],
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true
  },
  "exclude": ["node_modules/**"],
  "include": ["src"]
}
```

4. run typechecks: `npm run typecheck`

### Expected behavior

I expect no compilation errors at this point.

The error seems to be caused due to `@docusaurus/theme-classic` package. Consider the base tsconfig.json (at `@tsconfig/docusaurus/tsconfig.json`):

```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Docusaurus v2",
  "docs": "https://v2.docusaurus.io/docs/typescript-support",
  
  "compilerOptions": {
    "allowJs": true,
    "esModuleInterop": true,
    "jsx": "react",
    "lib": ["DOM"],
    "noEmit": true,
    "noImplicitAny": false,
    "types": ["node", "@docusaurus/module-type-aliases", "@docusaurus/theme-classic"],
    "baseUrl": ".",
    "paths": {
      "@site/*": ["./*"]
    }
  }
}
```

It includes `@docusaurus/theme-classic` package which mixes type definitions with ts source code. With this config, `tsc` scans several ts files at `@docusaurus` packages. This should be avoided whatsoever, a project build should not be related to ts files in external packages, only ts type definition files should be scanned instead.

As an example, the following ts config:

```json
{
  "extends": "@tsconfig/docusaurus/tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "exactOptionalPropertyTypes": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["ES2021"],
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "types": ["@docusaurus/module-type-aliases"]
  },
  "exclude": ["node_modules/**"],
  "include": ["src"]
}
```

compiles with no errors.

### Actual behavior

The following errors are printed:

```
node_modules/@docusaurus/plugin-content-docs/src/client/docsClientUtils.ts:123:3 - error TS2375: Type '{ activeVersion: GlobalVersion | undefined; activeDoc: GlobalDoc | undefined; alternateDocVersions: Record<string, GlobalDoc>; }' is not assignable to type 'ActiveDocContext' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
  Types of property 'activeVersion' are incompatible.
    Type 'GlobalVersion | undefined' is not assignable to type 'GlobalVersion'.
      Type 'undefined' is not assignable to type 'GlobalVersion'.

123   return {
      ~~~~~~~~
124     activeVersion,
    ~~~~~~~~~~~~~~~~~~
... 
126     alternateDocVersions: alternateVersionDocs,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
127   };
    ~~~~

node_modules/@docusaurus/plugin-content-docs/src/client/docsClientUtils.ts:145:3 - error TS2375: Type '{ latestDocSuggestion: GlobalDoc | undefined; latestVersionSuggestion: GlobalVersion; }' is not assignable to type 'DocVersionSuggestions' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
  Types of property 'latestDocSuggestion' are incompatible.
    Type 'GlobalDoc | undefined' is not assignable to type 'GlobalDoc'.
      Type 'undefined' is not assignable to type 'GlobalDoc'.

145   return {latestDocSuggestion, latestVersionSuggestion: latestVersion};
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors.

```

### Your environment

- Public source code: -
- Public site URL: -
- Docusaurus version used: 2.0.0-beta.14
- Environment name and version (e.g. Chrome 89, Node.js 16.4): NodeJs 16 (not relevant)
- Operating system and version (e.g. Ubuntu 20.04.2 LTS): Ubuntu 20 (not relevant)

### Self-service

- [X] I'd be willing to fix this bug myself.

**EDIT**: I removed `noUncheckedIndexedAccess` options since it's not related to the issue

**Comment by Josh-Cena** [comment]
Interesting error, never saw it... In the meantime, you can use `skipLibCheck`

**Comment by notaphplover** [comment]
> Interesting error, never saw it... In the meantime, you can use `skipLibCheck`

Hi @Josh-Cena, I am already using `"skipLibCheck": true`. The following tsconfig:

```json
{
  "extends": "@tsconfig/docusaurus/tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "exactOptionalPropertyTypes": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["ES2021"],
    "noFallthroughCasesInSwitch": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noUncheckedIndexedAccess": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true
  },
  "exclude": ["node_modules/**"],
  "include": ["src"]
}
```
fails to build as documented in the `Steps to reproduce` section.

**Comment by Josh-Cena** [comment]
Hmm, I see👍 If you can have any suggestions on how to improve, PRs are welcome. I do believe we have these options enabled in our repo already

**Comment by notaphplover** [comment]
I've been digging in the source code in order to understand how types are resolved. After this process I assert the following statements (please feel free to fix any wrong statement):

1. `docusaurus-module-type-aliases` package contains several module declarations, including every `@theme/` module a docusaurus theme must include.
2. `docusaurus-theme-classic` package includes in its types several `@theme/` module declarations, some of them are not included in `docusaurus-module-type-aliases` package like `@theme/EditThisPage`
3. Since statement `1` and `2` are right, developers interested in extending `docusaurus-theme-classic` components should use `docusaurus-theme-classic` types.
4. Adding `"@docusaurus/theme-classic"` to `types` in a `tsconfig` file force `tsc` compile undesired ts modules, not only ts definition modules.
5. `docusaurus-theme-classic` package includes in its types **only** module declarations.

Keeping this in mind, I propose the following fix:

1. Add a new package `docusaurus-theme-classic-types` with "@docusaurus/theme-classic" modules declaration. As an example, the name of the npm package could be `@docusaurus/theme-classic-types`
2. Release a version of the package.
3. Request https://github.com/tsconfig/bases to update docusaurus tsconfig to include `@docusaurus/theme-classic-types` instead of `@docusaurus/theme-classic`.
4. Request https://github.com/tsconfig/bases to release a new version of the package.
5. [BC, optional] remove module declaration types from `@docusaurus/theme-classic`.

This way tsc would only compile desired ts definitions instead of ts source code.

**Comment by Josh-Cena** [comment]
> * `docusaurus-module-type-aliases` package contains several module declarations, including every `@theme/` module a docusaurus theme must include.

This is incorrect. The `module-type-aliases` includes those Webpack aliases that the _core_ is aware of, so the `@theme/` alias only includes the `theme-fallback` components (5 of them).

> 2\. `docusaurus-theme-classic` package includes in its types several `@theme/` module declarations, some of them are not included in `docusaurus-module-type-aliases` package like `@theme/EditThisPage`

_Most_ of them are not included in `docusaurus-module-type-aliases`, because types here are for the aliases that only `theme-classic` provides. For example, we don't guarantee that another theme would also implement `@theme/EditThisPage`. `@theme/Layout` happens to be included in both declarations, because theme-classic re-implements the Layout to provide more styling.

> 3\. Since statement `1` and `2` are right, developers interested in extending `docusaurus-theme-classic` components should use `docusaurus-theme-classic` types.

Luckily, your conclusion is correct, that developers should include both declaration files.

> 4\. Adding `"@docusaurus/theme-classic"` to `types` in a `tsconfig` file force `tsc` compile undesired ts modules, not only ts definition modules.

This is the part that I don't understand. I thought the `types` field would only affect how _declarations_ are resolved, i.e. this is equivalent to using `<reference types="@docusaurus/theme-classic />` in your own declaration file.

> 5\. `docusaurus-theme-classic` package includes in its types **only** module declarations.

That's intentional. This declaration has to ambient because we are only interested in helping tsc resolve theme aliases.

---

Now, your solution is basically that we should release another types-only package like we did for `@docusaurus/theme-classic`. While that could be a way out, I don't like it from an architectural perspective because I want a package to be self-sustained. Note, though, that your errors are coming from `src/client/docsClientUtils.ts`, which is because we have a bad setup of type declaration files for these: the ambient declarations use `import('./index.ts')` to import types from actual TS files😅 This causes the `clientUtils` file to be compiled as well.

I think the actual solution should be to improve our declaration setup, and let the TS files import from .d.ts, not the other way round. We have a lot of bad legacy setup before I started working on this project, and I would try my best to fix it.

As a temporary solution, I realized that your error is because you have turned on `exactOptionalPropertyTypes`, turning it off makes the said errors be gone. FYI, our codebase is not ready for `exactOptionalPropertyTypes` yet (I checked that out before 4.4 was even released).

**Comment by notaphplover** [comment]
Thank you so much for those clarifications :).

> Note, though, that your errors are coming from `src/client/docsClientUtils.ts`, which is because we have a bad setup of type declaration files for these: the ambient declarations use `import('./index.ts')` to import types from actual TS files This causes the `clientUtils` file to be compiled as well.

I did not expect this to happen :smile:, ok now everything makes sense.

> As a temporary solution, I realized that your error is because you have turned on exactOptionalPropertyTypes, turning it off makes the said errors be gone. FYI, our codebase is not ready for exactOptionalPropertyTypes yet (I checked that out before 4.4 was even released).

Yeah, thats right, the error is caused due to `exactOptionalPropertyTypes` option. I can deal with it but I would hazard to say the issue should remain open until build does not break. As you said, type definitions should not include ts files. Third party typescript code should not have impact on a ts project and its not the case with these definitions.

> I think the actual solution should be to improve our declaration setup, and let the TS files import from .d.ts, not the other way round. We have a lot of bad legacy setup before I started working on this project, and I would try my best to fix it.

After reading your comment I totally agree with you. Regarding the use of `import('./index.ts')`, why don't you let typescript generate types and you import `import('./index.d.ts')` instead? That could deal with the issue.

**Comment by Josh-Cena** [comment]
The declaration files are not copied—they remain in the src folder. Also, on first build, the `index.d.ts` file doesn't exist yet, causing the `plugin-content-docs.d.ts` to fail to resolve. We have a terrible setup, I will improve soon.

> Third party typescript code should not have impact on a ts project and its not the case with these definitions.

Totally agree. I think we are on the same page.

**Comment by notaphplover** [comment]
Great! Ping me if you need a hand :)

**Comment by Josh-Cena** [comment]
@notaphplover Should be fixed by #6300. I still plan to refactor our type definitions further, but please test with canary release and tell me if it's fixed.

**Comment by notaphplover** [comment]
> @notaphplover Should be fixed by #6300. I still plan to refactor our type definitions further, but please test with canary release and tell me if it's fixed.

Sure, on my way to have lunch but I will test it before the day ends (here at GMT +1)

**Comment by Josh-Cena** [comment]
Going to sleep soon, but if it doesn't work out, I'll fix it tomorrow alongside #6175

**Comment by notaphplover** [comment]
@Josh-Cena with the canary build (0.0.0-4383) it works as expected! Thank you so much :)
